llama-util.h

```diff
@@ -219,7 +219,7 @@ struct llama_mmap {
         // prefetch/readahead impairs performance on NUMA systems
         if (numa) { prefetch = 0; }
 #ifdef __linux__
-        if (prefetch) { flags |= MAP_POPULATE; }
+        if (prefetch >= file->size) { flags |= MAP_POPULATE; }
```

Comment:
`prefetch` includes only the data of the tensors, without the headers. So I don't see how this could ever be true.
https://github.com/ggerganov/llama.cpp/blob/86c32198954a2bc482025703d6875e11f1c2a574/llama.cpp#L752-L757

Reply:
Thank you, I didn't think of that. In any case, if this is incorrect, then the code for determining the size for `madvise` is also incorrect:

https://github.com/ggerganov/llama.cpp/blob/86c32198954a2bc482025703d6875e11f1c2a574/llama-util.h#L229-L235

Speaking of which, doesn't this do the same thing as `MAP_POPULATE`? So couldn't we just remove `MAP_POPULATE` entirely?

Reply:
This part is a bit of a me, I am not really sure why all of that is needed, but I believe that when it was added it solved a problem for someone.
I am not sure how much it helps in practice, but in the backends interface PR I did some experimenting adding a `discard` method (`madvise` with `MADV_DONTNEED`) to suggest to the OS that the data won't be needed anymore after it is uploaded to the GPU. My reasoning is that the data will be needed, even if ultimately we upload it to the GPU, so populate and prefetch are fine, but what we can do is tell the OS that it can free these pages after we are done with them. 

https://github.com/ggerganov/llama.cpp/blob/d273bfd2c98148fab309f978c44ec2e8c24d1c4d/llama-util.h#L206

Reply:
So according to the `madvise` man page:

```
       MADV_WILLNEED
              Expect access in the near future.  (Hence, it might be a good idea to read some pages ahead.)
```

```
       MADV_POPULATE_READ (since Linux 5.14)
              "Populate (prefault) page tables readable, faulting in all pages in the range just as if manually reading from each page; however,
              avoid the actual memory access that would have been performed after handling the fault.

              In contrast to MAP_POPULATE, MADV_POPULATE_READ does not hide errors, can be applied to (parts of) existing mappings and will  al‐
              ways  populate (prefault) page tables readable.  One example use case is prefaulting a file mapping, reading all file content from
              disk; however, pages won't be dirtied and consequently won't have to be written back to disk when evicting the pages from memory.

              Depending on the underlying mapping, map the shared zeropage, preallocate memory or read the underlying  file;  files  with  holes
              might or might not preallocate blocks.  If populating fails, a SIGBUS signal is not generated; instead, an error is returned.

              If MADV_POPULATE_READ succeeds, all page tables have been populated (prefaulted) readable once.  If MADV_POPULATE_READ fails, some
              page tables might have been populated.

              MADV_POPULATE_READ  cannot be applied to mappings without read permissions and special mappings, for example, mappings marked with
              kernel-internal flags such as VM_PFNMAP or VM_IO, or secret memory regions created using memfd_secret(2).

              Note that with MADV_POPULATE_READ, the process can be killed at any moment when the system runs out of memory.
```

So it seems that `MADV_WILLNEED` will not necessarily prefetch the specified memory range but `MADV_POPULATE_READ` (Linux 5.14+ only) will.

Reply:
I did some testing: the discrepancy between the calculated prefetch size and the actual file size is ~500 kiB with only a very small increase as the model size increases. Meanwhile the size of a single 7b q2_K layer is ~80 MiB and the default size of a Linux memory page is 2 MiB. So we could do a kind of hacky fix where we just add something like 4 MiB to the prefetch size?

Reply:
Adding the size of the headers to `prefetch_size` should be easy and won't fail with different model sizes. You can assume that the padding is always 32 to simplify the code, it doesn't matter if the size is overestimated a bit.

>the default size of a Linux memory page is 2 MiB

This seems to be the size of huge pages, which I don't think we are using right now (but we probably should). Normal pages are 4kb.

Reply:
I pushed a new version with the following approach: start with the file size as the prefetch size and subtract from it the size of the offloaded tensors (which are at the end of the file). That way, when no tensors are offloaded the prefetch size is guaranteed to be the file size.

>This seems to be the size of huge pages, which I don't think we are using right now (but we probably should). Normal pages are 4kb.

On most Linux systems huge pages are used by default:

```
       MADV_HUGEPAGE (since Linux 2.6.38)
              Enable Transparent Huge Pages (THP) for pages in the range specified by addr and length.  The kernel will regularly scan the areas
              marked as huge page candidates to replace them with huge pages.  The kernel will also allocate huge pages directly when the region
              is naturally aligned to the huge page size (see posix_memalign(2)).

              This feature is primarily aimed at applications that use large mappings of data and access large regions of that memory at a  time
              (e.g., virtualization systems such as QEMU).  It can very easily waste memory (e.g., a 2 MB mapping that only ever accesses 1 byte
              will  result  in  2 MB  of  wired  memory  instead  of  one  4 KB  page).   See  the  Linux  kernel  source file Documentation/ad‐
              min-guide/mm/transhuge.rst for more details.

              Most common kernels configurations provide MADV_HUGEPAGE-style behavior by default, and thus MADV_HUGEPAGE is normally not  neces‐
              sary.  It is mostly intended for embedded systems, where MADV_HUGEPAGE-style behavior may not be enabled by default in the kernel.
              On such systems, this flag can be used in order to selectively enable THP.  Whenever MADV_HUGEPAGE is used, it should always be in
              regions  of  memory  with an access pattern that the developer knows in advance won't risk to increase the memory footprint of the
              application when transparent hugepages are enabled.

              Since Linux 5.4, automatic scan of eligible areas and replacement by huge pages works with private anonymous pages (see  mmap(2)),
              shmem  pages,  and  file-backed pages.  For all memory types, memory may only be replaced by huge pages on hugepage-aligned bound‐
              aries.  For file-mapped memory —including tmpfs (see tmpfs(2))— the mapping must also be  naturally  hugepage-aligned  within  the
              file.  Additionally, for file-backed, non-tmpfs memory, the file must not be open for write and the mapping must be executable.

              The  VMA must not be marked VM_NOHUGEPAGE, VM_HUGETLB, VM_IO, VM_DONTEXPAND, VM_MIXEDMAP, or VM_PFNMAP, nor can it be stack memory
              or backed by a DAX-enabled device (unless the DAX device  is  hot-plugged  as  System  RAM).   The  process  must  also  not  have
              PR_SET_THP_DISABLE set (see prctl(2)).

              The  MADV_HUGEPAGE,  MADV_NOHUGEPAGE,  and  MADV_COLLAPSE  operations  are  available  only if the kernel was configured with CON‐
              FIG_TRANSPARENT_HUGEPAGE and file/shmem memory is only supported if the kernel was configured with CONFIG_READ_ONLY_THP_FOR_FS.

```

Reply:
That should also work.

>On most Linux systems huge pages are used by default:

According to the [kernel documentation](https://www.kernel.org/doc/html/latest/admin-guide/mm/transhuge.html), the default is to require `madvise(MADV_HUGEPAGE)`, however in the Linux systems that I have access to, `cat /sys/kernel/mm/transparent_hugepage/enabled` returns `always`, but they are VMs, I don't know if the actual default is different under native Linux. We should probably add the `madvise(MADV_HUGEPAGE)` to be sure though, I don't think that there is any downside for us.